File: 310. Minimum Height Trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        graph = collections.defaultdict(list)
        for e in edges:
            graph[e[0]].append(e[1])
            graph[e[1]].append(e[0])
        
        leaves = [i for i in graph if len(graph[i])==1]
        
        while n > 2:
            n -= len(leaves)
            newlef = []
            for l in leaves:
                j = graph[l].pop()
                graph[j].remove(l)
                if len(graph[j]) == 1:
                    newlef.append(j)
            leaves = newlef
        return leaves

File: 310. Minimum Height Trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        HashMap<Integer, List<Integer>> graph = new HashMap<>();
        for (int i=0; i<n; i++) {
            graph.put(i, new ArrayList<>());
        }
        for (int[] e : edges) {
            graph.get(e[0]).add(e[1]);
            graph.get(e[1]).add(e[0]);
        }
        Deque<Integer> dq = new ArrayDeque<>();
        for (int i=0; i<n; i++) {
            if (graph.get(i).size() == 1) {
                dq.offer(i);
            }
        }
        while (graph.size() > 2) {
            int size = dq.size();
            for (int i=0; i<size; i++) {
                int curr = dq.poll();
                int nei = graph.get(curr).get(0);
                graph.get(nei).remove(new Integer(curr));
                if (graph.get(nei).size() == 1) {
                    dq.offer(nei);
                }
                graph.remove(curr);
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (int i : graph.keySet()) {
            ans.add(i);
        }
        return ans;
    }
}